File: Minimum Window Substring/answer.php
<?php
function findMin(string $txt, string $pattern): string{ 

    $tgt = array();
    foreach (str_split($pattern) as $ch) {
        if (!isset($tgt[$ch])) {
            $tgt[$ch] = 0;
        }
        $tgt[$ch]++;
    }

    $src = array();
    $best = "";
    $min_len = strlen($txt) + 1;
    $j = 0;

    for ($i = 0; $i < strlen($txt); $i++) {

        while ($j < strlen($txt) && !allCharsPresent($src, $tgt)) {
            $ch = $txt[$j];
            if (!isset($src[$ch])) {
                $src[$ch] = 0;
            }
            $src[$ch]++;
            $j++;
        }
        if (allCharsPresent($src, $tgt) && ($j - $i) < $min_len) {
            $min_len = $j - $i;
            $best = substr($txt, $i, $j - $i);
        }
        $src[$txt[$i]]--;
    }
    return $best;
}

function allCharsPresent($src, $tgt) {
    foreach ($tgt as $ch => $count) {
        if (!isset($src[$ch]) || $src[$ch] < $count) {
            return false;
        }
    }
    return true;
}

echo(findMin("ADOBECODEBANC" , "ABC"));

?>

File: Minimum Window Substring/answer.py
from collections import defaultdict
def findMin(txt:str , pattern:str) -> str:
    
    if not txt or not pattern or len(pattern) > len(txt):
        return ""
    
    src = defaultdict(int)
    tgt = defaultdict(int)

    for ch in pattern:
        tgt[ch] += 1

    best = ""
    min_len = len(txt) + 1
    right = 0

    for left in range(len(txt)):
        # Expand the window until it contains all required characters
        while right < len(txt) and not all(src.get(ch, 0) >= tgt[ch] for ch in tgt):
            src[txt[right]] += 1
            right += 1
            
        # If the window is valid and smaller than previously found, update best window
        if all(src.get(ch, 0) >= tgt[ch] for ch in tgt) and (right - left) < min_len:
            min_len = right - left
            best = txt[left:right]
        src[txt[left]] -= 1

    return(best)

print(findMin('ADOBECODEBANCE' , 'ABC')) 
